midend/simplifyBitwise.cpp

```diff
@@ -0,0 +1,128 @@
+#include "simplifyBitwise.h"
+
+namespace P4 {
+
+bool SimplifyBitwise::Scan::preorder(const IR::Operation *op) {
+    LOG1("Are we here? " << op);
+    can_be_changed = false;
+    return false;
+}
+
+bool SimplifyBitwise::Scan::preorder(const IR::AssignmentStatement *as) {
+    if (!as->right->is<IR::BOr>())
+        return false;
+    auto action = findContext<IR::P4Action>();
+    if (action)
+        LOG1("Action name " << action->name.name);
+    can_be_changed = true;
+    total_mask = 0;
+    return true;
+}
+
+bool SimplifyBitwise::Scan::preorder(const IR::BOr *bor) {
+    LOG1("Bor");
+    if (!findContext<IR::AssignmentStatement>())
+        can_be_changed = false;
+    if (!(bor->left->is<IR::BAnd>() && bor->right->is<IR::BAnd>()))
+        can_be_changed = false;
+    LOG1("Tests " << bor->left->is<IR::BAnd>() << " " << bor->right->is<IR::BAnd>());
+    return can_be_changed;
+}
+
+bool SimplifyBitwise::Scan::preorder(const IR::BAnd *band) {
+    LOG1("Band");
+    if (!findContext<IR::BOr>())
+        can_be_changed = false;
+    if (!band->left->is<IR::Constant>() && !band->right->is<IR::Constant>())
+        can_be_changed = false;
+    if (band->left->is<IR::Constant>() && band->right->is<IR::Constant>())
+        can_be_changed = false;
+    return can_be_changed;
+}
+
+bool SimplifyBitwise::Scan::preorder(const IR::Constant *constant) {
+    LOG1("Constant");
+    if (!findContext<IR::BAnd>()) {
+        can_be_changed = false;
+        return false;
+    }
+    // Ensure that there are no collisions within the bitmask
+    mpz_class mask = 0;
+    mask = total_mask & constant->value;
+    if (mask == 0) {
+        total_mask |= constant->value;
+    } else {
+        can_be_changed = false;
+    }
+    return false;
+}
+
+void SimplifyBitwise::Scan::postorder(const IR::AssignmentStatement *as) {
+    if (can_be_changed)
+        self.changeable.insert(as);
+}
+
+const IR::Node *SimplifyBitwise::Update::preorder(IR::AssignmentStatement *as) {
+    auto orig_as = getOriginal()->to<IR::AssignmentStatement>();
```

Comment:
As an aside, you can use `getOriginal<IR::AssignmentStatement>()` to do the getOriginal and dynamic cast and BUG_CHECK if its not of the expected type, all in one step (as this is such a common idiom).